### PR TITLE
@semanticNonNull support

### PR DESCRIPTION
**Pull request by tobias-tengler**
Adds a new `EnableSemanticNonNull` option that, when set to `true`, rewrites all non-null fields to nullable fields and annotates them with the `@semanticNonNull` directive. A ResultFormatter is added to the rewritten fields that throws an error if a `null` value is returned at a position that is marked as semantically non-null.

```csharp
builder.Services
  .AddGraphQLServer()
  .ModifyOptions(o =>
  {
    o.EnableSemanticNonNull = true;
  });
```

The `@semanticNonNull` directive has a `levels` argument that allows to specify the semantic non-nullability in the context of lists. `0` is the outer most list and _n_ the named type.

|Non-Null|Semantic Non-Null|
|---|---|
|`field: String!`|`field: String @semanticNonNull`|
|`field: [String!]`|`field: [String] @semanticNonNull(levels: [ 1 ])`|
|`field: [[String!]]!`|`field: [[String]] @semanticNonNull(levels: [ 0, 2 ])`|

### CODE REVIEW

File: src/HotChocolate/Core/src/Types/SchemaOptions.cs [hunk 1]
```diff
@@ -15,58 +15,34 @@ public class SchemaOptions : IReadOnlySchemaOptions
     private BindingBehavior _defaultBindingBehavior = BindingBehavior.Implicit;
     private FieldBindingFlags _defaultFieldBindingFlags = FieldBindingFlags.Instance;
 
-    /// <summary>
-    /// Gets or sets the name of the query type.
-    /// </summary>
+    /// <inheritdoc cref="IReadOnlySchemaOptions.QueryTypeName"/>
```
Comment (glen-84) [inline src/HotChocolate/Core/src/Types/SchemaOptions.cs:18]: I don't think that these crefs are needed.

File: src/HotChocolate/Core/test/Types.Tests/SemanticNonNullTests.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            .AddGraphQL()
+            .ModifyOptions(o => o.EnableSemanticNonNull = true)
+            .AddQueryType<Query>()
+            .BuildSchemaAsync()
+            .MatchSnapshotAsync();
+    }
+
+     public class Query
+    {
+        [GraphQLNonNullType()]
```
Comment (glen-84) [inline src/HotChocolate/Core/test/Types.Tests/SemanticNonNullTests.cs:24]:
```suggestion
        [GraphQLNonNullType]
```

File: src/HotChocolate/Core/test/Execution.Tests/SemanticNonNullTests.cs [hunk 3]
```diff
@@ -0,0 +1,1052 @@
...
+        }
+
+        public string[] PureScalarListItemReturningNull => ["a", null!, "c"];
+
+        public string[] PureScalarListItemThrowingError(IResolverContext context)
+        {
+            // TODO: How can you create a terminating error for a single item?
+            context.ReportError(ErrorBuilder.New().SetMessage("Another error").SetPath(context.Path.Append(1)).Build());
+            return ["a", null!, "c"];
+        }
```
Comment (tobias-tengler) [inline src/HotChocolate/Core/test/Execution.Tests/SemanticNonNullTests.cs:943]: Does someone have an idea?

File: src/HotChocolate/Core/src/Types/SchemaOptions.cs [hunk 4]
```diff
@@ -15,58 +15,24 @@ public class SchemaOptions : IReadOnlySchemaOptions
     private BindingBehavior _defaultBindingBehavior = BindingBehavior.Implicit;
     private FieldBindingFlags _defaultFieldBindingFlags = FieldBindingFlags.Instance;
 
-    /// <summary>
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/SchemaOptions.cs:18]: why did you remove the xml docs?

Comment (tobias-tengler) [reply src/HotChocolate/Core/src/Types/SchemaOptions.cs:18]: It's not needed. It's provided through the underlying interface. This way the descriptions only need to be maintained on the interface.

Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/SchemaOptions.cs:18]: All public types must have xml docs even with interface. But you can add an inheritance xml tag

Comment (tobias-tengler) [reply src/HotChocolate/Core/src/Types/SchemaOptions.cs:18]: Where does this must come from? Is it convention of this code base, because Rider and VS Code don't require it.

Comment (tobias-tengler) [reply src/HotChocolate/Core/src/Types/SchemaOptions.cs:18]: I've added inheritdoc regardless for now...
